client/navigation/utils.ts

```diff
@@ -67,20 +72,33 @@ export const getDefaultMatchExpression = ( url ) => {
 	return '^' + path + argsExpression + hashExpression;
 };
 
+interface wcNavigation {
+	menuItems: Array< Item >;
+	rootBackLabel: string;
+	rootBackUrl: string;
+	historyPatched: boolean;
+}
+
+declare global {
+	interface Window {
+		wcNavigation: wcNavigation;
+	}
+}
+
 /**
  * Adds a listener that runs on history change.
  *
  * @param {Function} listener Listener to add on history change.
  * @return {Function} Function to remove listeners.
  */
-export const addHistoryListener = ( listener ) => {
+export const addHistoryListener = ( listener: Function ) => {
 	// Monkey patch pushState to allow trigger the pushstate event listener.
 	if ( ! window.wcNavigation.historyPatched ) {
 		( ( history ) => {
 			/* global CustomEvent */
 			const pushState = history.pushState;
 			const replaceState = history.replaceState;
-			history.pushState = function ( state ) {
+			history.pushState = function ( state: object ) {
```

Comment:
I believe `object` types out to be:
```
{ [key: string]: any }
```
Which is in a way a sneaky way to use `any` ;) Would you know at-least some of the properties here that you could create an interface for it? Even if all the key's are optional for now: `?:`